File: src/ViewModel/CardViewModel.cpp
#include "CardViewModel.h"
#include "Card.h"

CardViewModel::CardViewModel(Card* card, QObject* parent)
    : QObject(parent)
    , m_card(card)
{
}

int CardViewModel::id() const { return m_card ? m_card->id() : -1; }
CardType CardViewModel::cardType() const { return m_card ? m_card->cardType() : CardType::Kill; }
CardSuit CardViewModel::suit() const { return m_card ? m_card->suit() : CardSuit::Spade; }
int CardViewModel::number() const { return m_card ? m_card->number() : 0; }
QString CardViewModel::cardName() const { return m_card ? QString::fromStdString(m_card->cardName()) : QString(); }
QString CardViewModel::description() const { return m_card ? QString::fromStdString(m_card->description()) : QString(); }
CardColor CardViewModel::color() const { return m_card ? m_card->color() : CardColor::Black; }
bool CardViewModel::isBasic() const { return m_card ? m_card->isBasic() : false; }
bool CardViewModel::isStrategy() const { return m_card ? m_card->isStrategy() : false; }
QString CardViewModel::suitSymbol() const { return m_card ? QString::fromStdString(m_card->suitSymbol()) : QString(); }
QString CardViewModel::numberString() const { return m_card ? QString::fromStdString(m_card->numberString()) : QString(); }
Card* CardViewModel::card() const { return m_card; }

// ==================== UI 状态 ====================

bool CardViewModel::isSelected() const { return m_selected; }

void CardViewModel::setSelected(bool selected)
{
    if (m_selected != selected) {
        m_selected = selected;
        emit selectedChanged(selected);
    }
}

bool CardViewModel::isPlayable() const { return m_playable; }

void CardViewModel::setPlayable(bool playable)
{
    if (m_playable != playable) {
        m_playable = playable;
        emit playableChanged(playable);
    }
}

bool CardViewModel::isHighlighted() const { return m_highlighted; }

void CardViewModel::setHighlighted(bool highlighted)
{
    if (m_highlighted != highlighted) {
        m_highlighted = highlighted;
        emit highlightedChanged(highlighted);
    }
}

void CardViewModel::resetUIState()
{
    setSelected(false);
    setPlayable(false);
    setHighlighted(false);
}

QString CardViewModel::cardTypeName(CardType type)
{
    return QString::fromStdString(Card::cardTypeName(type));
}

File: Unity/Scripts/ViewModel/CardViewModel.cs
using System.Collections.Generic;

/// <summary>
/// 卡牌 ViewModel — 在 Card 基础上增加 UI 展示状态（是否选中、是否可打出、是否高亮）
/// </summary>
public class CardViewModel
{
    private readonly Card _card;
    private bool _selected = false;
    private bool _playable = false;
    private bool _highlighted = false;

    public CardViewModel(Card card)
    {
        _card = card;
    }

    // ==================== Card 只读属性（透传）====================

    public int Id => _card?.Id ?? -1;
    public CardType CardType => _card?.Type ?? CardType.Kill;
    public CardSuit Suit => _card?.Suit ?? CardSuit.Spade;
    public int Number => _card?.Number ?? 0;
    public string CardName => _card?.CardName ?? "";
    public string Description => _card?.Description ?? "";
    public CardColor Color => _card?.Color ?? CardColor.Black;
    public bool IsBasic => _card?.IsBasic ?? false;
    public bool IsStrategy => _card?.IsStrategy ?? false;

    public string SuitSymbol()
    {
        return _card?.SuitSymbol().ToString() ?? "?";
    }

    public string NumberString()
    {
        return _card?.NumberString() ?? "0";
    }

    // ==================== UI 状态 ====================

    public bool IsSelected => _selected;

    public void SetSelected(bool selected)
    {
        if (_selected != selected)
        {
            _selected = selected;
            SelectedChanged?.Emit(selected);
        }
    }

    public bool IsPlayable => _playable;

    public void SetPlayable(bool playable)
    {
        if (_playable != playable)
        {
            _playable = playable;
            PlayableChanged?.Emit(playable);
        }
    }

    public bool IsHighlighted => _highlighted;

    public void SetHighlighted(bool highlighted)
    {
        if (_highlighted != highlighted)
        {
            _highlighted = highlighted;
            HighlightedChanged?.Emit(highlighted);
        }
    }

    /// <summary>重置所有 UI 状态</summary>
    public void ResetUIState()
    {
        SetSelected(false);
        SetPlayable(false);
        SetHighlighted(false);
    }

    // ==================== 原始指针访问 ====================

    public Card Card => _card;

    // ==================== 事件 ====================

    public readonly Signal<bool> SelectedChanged = new();
    public readonly Signal<bool> PlayableChanged = new();
    public readonly Signal<bool> HighlightedChanged = new();
}
